Sources/OpenCombine/Publishers/Publishers.ReplaceEmpty.swift

```diff
@@ -0,0 +1,177 @@
+#endif
+
+extension Publisher {
+
+    /// Replaces an empty stream with the provided element.
+    ///
+    /// If the upstream publisher finishes without producing any elements,
+    /// this publisher emits the provided element, then finishes normally.
+    /// - Parameter output: An element to emit when the upstream publisher
+    ///                     finishes without emitting any elements.
+    /// - Returns: A publisher that replaces an empty stream with
+    ///            the provided output element.
+    public func replaceEmpty(with output: Output) -> Publishers.ReplaceEmpty<Self> {
+        return .init(upstream: self, output: output)
+    }
+}
+
+extension Publishers {
+
+    /// A publisher that replaces an empty stream with a provided element.
+    public struct ReplaceEmpty<Upstream>: Publisher where Upstream: Publisher {
+
+        /// The kind of values published by this publisher.
+        public typealias Output = Upstream.Output
+
+        /// The kind of errors this publisher might publish.
+        ///
+        /// Use `Never` if this `Publisher` does not publish errors.
+        public typealias Failure = Upstream.Failure
+
+        /// The element to deliver when the upstream publisher finishes
+        /// without delivering any elements.
+        public let output: Upstream.Output
+
+        /// The publisher from which this publisher receives elements.
+        public let upstream: Upstream
+
+        public init(upstream: Upstream, output: Output) {
+            self.upstream = upstream
+            self.output = output
+        }
+
+        /// This function is called to attach the specified `Subscriber`
+        /// to this `Publisher` by `subscribe(_:)`
+        ///
+        /// - SeeAlso: `subscribe(_:)`
+        /// - Parameters:
+        ///     - subscriber: The subscriber to attach to this `Publisher`.
+        ///                   once attached it can begin to receive values.
+        public func receive<Downstream: Subscriber>(subscriber: Downstream)
+            where Upstream.Failure == Downstream.Failure,
+                  Upstream.Output == Downstream.Input
+        {
+            let inner = Inner(downstream: subscriber, output: output)
+            upstream.subscribe(inner)
+        }
+    }
+}
+
+extension Publishers.ReplaceEmpty {
+
+    private final class Inner<Downstream: Subscriber>
+        : Subscriber,
+          Subscription,
+          CustomStringConvertible,
+          CustomReflectable,
+          CustomPlaygroundDisplayConvertible
+        where Upstream.Failure == Downstream.Failure,
+              Upstream.Output == Downstream.Input
+    {
+
+        typealias Input = Upstream.Output
+        typealias Failure = Upstream.Failure
+
+        private let output: Output
+        private let downstream: Downstream
+
+        private var status = SubscriptionStatus.awaitingSubscription
+        private var lock = UnfairLock.allocate()
+        private var isEmpty = true
```

Comment:
Using [SuperMirror](https://github.com/broadwaylamb/SuperMirror) I managed to look at all the fields that Combine has here:

- `output: Output`
- `downstream: Downstream`
- `receivedUpstream: Bool` (`isEmpty` in your implementation)
- `lock: UnfairLock`
- `downstreamRequested: Bool`
- `finishedWithoutUpstream: Bool`
- `status: SubscriptionStatus`

1. `downstreamRequested` is used to decide whether to send the extra value downstream. If the upstream has finished without producing a value, **and** downstream hasn't requested anything yet, we must not send `output` downstream, we rather should wait for a request. Could you please add a test for it and tweak your implementation for the test to pass?

2. `finishedWithoutUpstream` seems to be set to `true` when we receive completion from the upstream, but don't have any demand from the downstream yet.

3. Can I ask you to rename the existing stored properties so that they match Combine's names, and also add and leverage those flags that I mentioned? I'd also be good to preserve the order of the stored properties. Why? First of all, we want to stay as close to Combine as possible implementation-wise. Also, the other goal of OpenCombine that I have in mind is that it should be a good reference for the people curious about how Combine works. Third, by speaking in the same "language" as Combine, we can make debugging easier: if something works in Combine and doesn't work in OpenCombine, it'd be good if we could inspect state of Combine objects (the values of their stored properties) and immediately map that state to OpenCombine's to see what went wrong.

Reply:
@broadwaylamb I'm think I've got 1 and 3 covered, but I'm not sure if there is a valid test that we could use for 2.  Let me know if you have any suggestions, but otherwise I've renamed and added everything.